crates/sui-storage/src/write_ahead_log.rs

```diff
@@ -173,27 +186,55 @@ where
 
         Self {
             log,
+            retry_count,
             recoverable_txes: Mutex::new(recoverable_txes),
             mutex_table: MutexTable::new(MUTEX_TABLE_SIZE),
         }
     }
 
     fn commit_tx(&self, tx: &TransactionDigest) -> SuiResult {
         debug!(digest = ?tx, "committing tx");
-        self.log.remove(tx).map_err(|e| e.into())
+        let write_batch = self.log.batch();
+        let write_batch = write_batch.delete_batch(&self.log, std::iter::once(tx))?;
+        let write_batch = write_batch.delete_batch(&self.retry_count, std::iter::once(tx))?;
+        write_batch.write().map_err(SuiError::from)
+    }
+
+    fn increment_retry_count(&self, tx: &TransactionDigest) -> SuiResult {
+        let cur = self.retry_count.get(tx)?.unwrap_or(0);
+        self.retry_count
+            .insert(tx, &(cur + 1))
+            .map_err(SuiError::from)
     }
 
     fn implicit_drop_tx(&self, tx: &TransactionDigest) {
         // this function should be called very rarely so contention should not be an issue.
         // unwrap ok because it is not safe to continue running if the mutex is poisoned.
-        let mut r = self.recoverable_txes.lock().unwrap();
-        r.push(*tx);
+        self.recoverable_txes.lock().unwrap().push(*tx);
     }
 
     fn pop_one_tx(&self) -> Option<TransactionDigest> {
         // Only acquire this lock inside a sync function to make sure we don't accidentally
-        // hold it across an .await
-        self.recoverable_txes.lock().unwrap().pop()
+        // hold it across an .await - unwrap okay because we should crash if a mutex is
+        // poisoned.
+        let recoverable_txes = &mut self.recoverable_txes.lock().unwrap();
+
+        while let Some(tx) = recoverable_txes.pop() {
+            if let Err(e) = self.increment_retry_count(&tx) {
```

Comment:
Do we ever delete entries from the `retry_count` table?

Reply:
Yes, see `commit_tx()`